### PR TITLE
[WIP] Add a quantum tape

### PR DESCRIPTION
**Pull request by josh146**

### CODE REVIEW

File: pennylane/beta/interfaces/autograd.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+    .. note::
+
+        If using a device that supports native autograd computation and backpropagation, such as
+        :class:`~.DefaultQubitAutograd`, the Autograd interface **does not need to be applied**. It
+        is only applied to tapes executed on non-Autograd compatible devices.
+
+    **Example**
+
+    One an autograd quantum tape has been created, it can be differentiated using autograd:
```
Comment (trbromley) [inline pennylane/beta/interfaces/autograd.py:51]:
```suggestion
    Once an autograd quantum tape has been created, it can be differentiated using autograd:
```

File: pennylane/beta/interfaces/autograd.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        params = [item for sublist in params for item in sublist]
+
+        trainable_params = set()
+
+        for idx, p in enumerate(params):
+            if getattr(p, "requires_grad", True):
+                trainable_params.add(idx)
+
+        self.trainable_params = trainable_params
+        return params
```
Comment (trbromley) [inline pennylane/beta/interfaces/autograd.py:92]: Should we be returning something and updating the trainable params? It feels like we should just repeat the two lines
```python
        params = [o.data for o in self.operations + self.observables]
        params = [item for sublist in params for item in sublist]
```
within `get_parameters()`.

File: pennylane/beta/interfaces/autograd.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+    def get_parameters(self, free_only=True):  # pylint: disable=missing-function-docstring
+        params = self._update_trainable_params()
+
+        if free_only:
+            params = [p for idx, p in enumerate(params) if idx in self.trainable_params]
```
Comment (trbromley) [inline pennylane/beta/interfaces/autograd.py:98]: Should we use `trainable_only` as the argument?

File: pennylane/beta/interfaces/autograd.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+        res = self.execute_device(params, device=device)
+
+        if res.dtype == np.dtype("object"):
+            return np.hstack(res)
+
+        return res
+
+    @staticmethod
+    def vjp(ans, self, params, device):  # pylint: disable=unused-argument
```
Comment (trbromley) [inline pennylane/beta/interfaces/autograd.py:114]:
```suggestion
    def vjp(self, ans, params, device):  # pylint: disable=unused-argument
```
Although I'm not sure what `ans` is :thinking:

File: pennylane/beta/interfaces/autograd.py (hunk 3 above)
Comment (josh146) [reply pennylane/beta/interfaces/autograd.py:98]: Yes, this is nicer than free_only!

File: pennylane/beta/interfaces/autograd.py (hunk 2 above)
Comment (josh146) [reply pennylane/beta/interfaces/autograd.py:92]: Yes, probably 😆 This was probably me not wanting to repeat those two lines, I have a fear of introducing bottlenecks as we build out the new core

File: pennylane/beta/interfaces/autograd.py (hunk 4 above)
Comment (josh146) [reply pennylane/beta/interfaces/autograd.py:114]: I'm a bit hesitant to change this; I think `ans` might be required to be the first argument by autograd

Comment (trbromley) [reply pennylane/beta/interfaces/autograd.py:114]: Ah I see!

File: pennylane/beta/interfaces/autograd.py (hunk 2 above)
Comment (josh146) [reply pennylane/beta/interfaces/autograd.py:92]: @trbromley interesting that `get_parameters()` is the next bottleneck after `_grad_method`. Are these two list comprehensions the culprit, or is there something else I am missing?

File: pennylane/beta/tapes/qubit_param_shift.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+                # check if op is an ancestor of ob
+                has_path = self.graph.has_path(op, ob)
+
+                # Use analytic method if there is a path, else the gradient is zero
+                best.append("A" if has_path else "0")
+
+            if all(k == "0" for k in best):
+                return "0"
+
+        return "A"
```
Comment (trbromley) [inline pennylane/beta/tapes/qubit_param_shift.py:80]: Could we make this have less duplication? e.g.
```python
def _grad_method(self, idx, use_graph=True):
    op = self._par_info[idx]["op"]

    if op.grad_method == "F":
            return "F"

    return super()._grad_method(idx, use_graph=use_graph, meth="A")
```

**Review by trbromley — APPROVED** [review]
Awesome @josh146! I'm approving this - my understanding from discussions is that we'll have smaller follow up PRs where we can go into more depth. For now, this looks great and my bar for approval was trialing out some code in the new approach.
:100:

**Comment by trbromley** [comment]
One thing I noticed while playing with the quantum tape:
This works nicely:
```python
import tensorflow as tf

import pennylane as qml

from pennylane.beta.tapes import QuantumTape
from pennylane.beta.queuing import expval
from pennylane.beta.interfaces.tf import TFInterface

x = tf.Variable(0.5, dtype=tf.float64)
dev = qml.device("default.qubit", wires=3)

tape = TFInterface.apply(QuantumTape())

with tf.GradientTape() as ctape:
    with tape:
        qml.RX(x, wires=0)
        qml.RX(2 * x ** 6, wires=1)
        qml.RX(tf.sin(x), wires=0)
        qml.CNOT(wires=[0, 1])
        qml.CNOT(wires=[2, 0])
        qml.RY(tf.exp(x), wires=2)
        expval(qml.PauliZ(0))
        expval(qml.PauliZ(1))
        expval(qml.PauliZ(2))
    res = tape.execute(dev)
    
grad = ctape.jacobian(res, x)
print(grad)
```

But, this gives the wrong result:
```python
import tensorflow as tf

import pennylane as qml

from pennylane.beta.tapes import QuantumTape
from pennylane.beta.queuing import expval
from pennylane.beta.interfaces.tf import TFInterface

x = tf.Variable(0.5, dtype=tf.float64)
dev = qml.device("default.qubit", wires=3)

tape = TFInterface.apply(QuantumTape())

with tape:
    qml.RX(x, wires=0)
    qml.RX(2 * x ** 6, wires=1)
    qml.RX(tf.sin(x), wires=0)
    qml.CNOT(wires=[0, 1])
    qml.CNOT(wires=[2, 0])
    qml.RY(tf.exp(x), wires=2)
    expval(qml.PauliZ(0))
    expval(qml.PauliZ(1))
    expval(qml.PauliZ(2))

with tf.GradientTape() as ctape:
    res = tape.execute(dev)
    
grad = ctape.jacobian(res, x)
print(grad)
```
That is, one must run the quantum tape within the classical tape. This probably makes sense, but I can see the second use case happening with users (not sure how much we'll advertise the tape approach vs the qnode approach though).

Also, would there ever be a case where we can avoid the `tf.GradientTape()` completely?
```python
import tensorflow as tf

import pennylane as qml

from pennylane.beta.tapes import QuantumTape
from pennylane.beta.queuing import expval
from pennylane.beta.interfaces.tf import TFInterface

x = tf.Variable(0.5, dtype=tf.float64)
dev = qml.device("default.qubit", wires=3)

tape = TFInterface.apply(QuantumTape())

with tape:
    qml.RX(x, wires=0)
    qml.RX(2 * x ** 6, wires=1)
    qml.RX(tf.sin(x), wires=0)
    qml.CNOT(wires=[0, 1])
    qml.CNOT(wires=[2, 0])
    qml.RY(tf.exp(x), wires=2)
    expval(qml.PauliZ(0))
    expval(qml.PauliZ(1))
    expval(qml.PauliZ(2))
    
grad = tape.jacobian(x, device=dev)
print(grad)
```
e.g., something like above.

**Comment by josh146** [comment]
While the following will give the correct result, I'm not surprised it gives the incorrect gradient!

```python
x = tf.Variable(0.5, dtype=tf.float64)

with TFInterface.apply(QuantumTape()) as qtape:
    qml.RX(2 * x ** 6, wires=1)
    qml.RX(tf.sin(x), wires=0)
    expval(qml.PauliZ(0))

with tf.GradientTape() as ctape:
    res = qtape.execute(dev)

grad = ctape.jacobian(...)
```
This is because the classical processing of the gate arguments (`2 * x ** 6`, `tf.sin(x)`, etc.) is happening _outside_ of the TensorFlow tape, so TensorFlow is not taking the classical processing into account when returning the Jacobian! It is returning only the pure quantum Jacobian, with respect to the quantum gate arguments (compare the result to calling `qtape.jacobian(dev)`).

If you want to differentiate wrt `x`, not the gate arguments, you **need** to include any intermediate classical processing of `x` within the classical gradient tape.

> Also, would there ever be a case where we can avoid the tf.GradientTape() completely?

Unfortunately, the gradient tape is a requirement if you want to perform differentiable classical processing at any point in the computation. We _could_ have the `TFQuantumTape` implicitly return a quantum tape _and_ a tf gradient tape, but I think this is attempting to do too much implicitly for the user. There are a lot of cases where you might _already_ have a pre-existing gradient tape, or you want to control the tape settings.

File: pennylane/beta/tapes/qubit_param_shift.py (hunk 5 above)
Comment (josh146) [reply pennylane/beta/tapes/qubit_param_shift.py:80]: good idea!

File: pennylane/beta/tapes/qnode.py [hunk 6]
```diff
@@ -0,0 +1,528 @@
...
+
+        stop_at = self.device.operations
+
+        if isinstance(self.qtape, QubitParamShiftTape):
+            # controlled rotations aren't supported by the parameter-shift rule
+            stop_at = set(self.device.operations) - {"CRX", "CRZ", "CRY", "CRot"}
+
+        # expand out the tape, if any operations are not supported on the device
+        if not {op.name for op in self.qtape.operations}.issubset(stop_at):
+            self.qtape = self.qtape.expand(depth=self.max_expansion, stop_at=stop_at)
```
Comment (josh146) [inline pennylane/beta/tapes/qnode.py:359]: @trbromley I temporarily added this. It expands out the tape to remove CRm gates if the tape is a parameter shift one.

File: tests/beta/interfaces/test_tape_autograd.py [hunk 7]
```diff
@@ -607,11 +667,21 @@ def cost(x, y, device):
...
+
+        def _asarray(args, dtype=np.float64):
+            return np.hstack(args).flatten()
+
+        # The current DefaultQubitAutograd device provides an _asarray method that does
+        # not work correctly for ragged arrays. For ragged arrays, we would like _asarray to
+        # flatten the array. Here, we patch the _asarray method on the device to achieve this
+        # behaviour; once the tape has moved from the beta folder, we should implement
+        # this change directly in the device.
+        monkeypatch.setattr(dev, "_asarray", _asarray)
```
Comment (co9olguy) [inline tests/beta/interfaces/test_tape_autograd.py]:
```suggestion
        # behaviour
        # TODO: once the tape has moved from the beta folder, we should implement
        # this change directly in the device.
        monkeypatch.setattr(dev, "_asarray", _asarray)

```

Comment (co9olguy) [reply tests/beta/interfaces/test_tape_autograd.py]: don't want to forget

File: tests/beta/interfaces/test_tape_tf.py [hunk 8]
```diff
@@ -697,7 +737,11 @@ def _asarray(args, dtype=tf.float64):
             res = tf.concat(res, axis=0)
             return tf.cast(res, dtype=dtype)
 
-        # we need to patch the asarray method on the device
+        # The current DefaultQubitTF device provides an _asarray method that does
+        # not work correctly for ragged arrays. For ragged arrays, we would like _asarray to
+        # flatten the array. Here, we patch the _asarray method on the device to achieve this
+        # behaviour; once the tape has moved from the beta folder, we should implement
+        # this change directly in the device.
```
Comment (co9olguy) [inline tests/beta/interfaces/test_tape_tf.py]:
```suggestion
        # behaviour
        # TODO: once the tape has moved from the beta folder, we should implement
        # this change directly in the device.

```

**Comment by co9olguy** [comment]
:crying_cat_face:

**Comment by josh146** [comment]
This was my original prototype branch :laughing:
